commons-core/src/main/scala/com/avsystem/commons/rest/data.scala

```diff
@@ -0,0 +1,195 @@
+    case HttpBody.Empty => Opt.Empty
+  }
+
+  def forNonEmpty(consumer: (String, String) => Unit): Unit = this match {
+    case HttpBody(content, mimeType) => consumer(content, mimeType)
+    case HttpBody.Empty =>
+  }
+
+  def readContent(): String = this match {
+    case HttpBody(content, _) => content
+    case HttpBody.Empty => throw new ReadFailure("Expected non-empty body")
+  }
+
+  def readJson(): JsonValue = this match {
+    case HttpBody(content, HttpBody.JsonType) => JsonValue(content)
+    case HttpBody(_, mimeType) =>
+      throw new ReadFailure(s"Expected body with application/json type, got $mimeType")
+    case HttpBody.Empty =>
+      throw new ReadFailure("Expected body with application/json type, got empty body")
+  }
+}
+object HttpBody {
+  object Empty extends HttpBody
+  final case class NonEmpty(content: String, mimeType: String) extends HttpBody
+
+  def empty: HttpBody = Empty
+
+  def apply(content: String, mimeType: String): HttpBody =
+    NonEmpty(content, mimeType)
+
+  def unapply(body: HttpBody): Opt[(String, String)] = body match {
+    case Empty => Opt.Empty
+    case NonEmpty(content, mimeType) => Opt((content, mimeType))
+  }
+
+  final val PlainType = "text/plain"
+  final val JsonType = "application/json"
+
+  def plain(value: String): HttpBody = HttpBody(value, PlainType)
+  def json(json: JsonValue): HttpBody = HttpBody(json.value, JsonType)
+
+  def createJsonBody(fields: NamedParams[JsonValue]): HttpBody =
+    if (fields.isEmpty) HttpBody.Empty else {
+      val sb = new JStringBuilder
+      val oo = new JsonStringOutput(sb).writeObject()
+      fields.foreach {
+        case (key, JsonValue(json)) =>
+          oo.writeField(key).writeRawJson(json)
+      }
+      oo.finish()
+      HttpBody.json(JsonValue(sb.toString))
+    }
+
+  def parseJsonBody(body: HttpBody): NamedParams[JsonValue] = body match {
+    case HttpBody.Empty => NamedParams.empty
+    case _ =>
+      val oi = new JsonStringInput(new JsonReader(body.readJson().value)).readObject()
+      val builder = NamedParams.newBuilder[JsonValue]
+      while (oi.hasNext) {
+        val fi = oi.nextField()
+        builder += ((fi.fieldName, JsonValue(fi.readRawJson())))
+      }
+      builder.result()
+  }
+
+  implicit val emptyBodyForUnit: AsRawReal[HttpBody, Unit] =
+    AsRawReal.create(_ => HttpBody.Empty, _ => ())
+  implicit def httpBodyJsonAsRaw[T](implicit jsonAsRaw: AsRaw[JsonValue, T]): AsRaw[HttpBody, T] =
+    AsRaw.create(v => HttpBody.json(jsonAsRaw.asRaw(v)))
+  implicit def httpBodyJsonAsReal[T](implicit jsonAsReal: AsReal[JsonValue, T]): AsReal[HttpBody, T] =
+    AsReal.create(v => jsonAsReal.asReal(v.readJson()))
+}
+
+/**
+  * Enum representing HTTP methods.
+  */
+final class HttpMethod(implicit enumCtx: EnumCtx) extends AbstractValueEnum
+object HttpMethod extends AbstractValueEnumCompanion[HttpMethod] {
+  final val GET, PUT, POST, PATCH, DELETE: Value = new HttpMethod
+}
+
+case class RestHeaders(
```

Comment:
This name is a little misleading - it contains not only headers, but also other arguments. What about `RestParameters`?

Reply:
`RestParameters` is not perfect either because there are also body parameters which are not part of this class. I'm open to other naming suggestions but I couldn't find better.

Reply:
I agree, `RestParameters` is not perfect, but it's a little better.